File: go/binary_search.go
package main

import "golang.org/x/exp/constraints"

// Ordering represents the result of a comparison: Less, Equal, or Greater
type Ordering int

const (
	Less Ordering = iota
	Equal
	Greater
)

func compare_function[T constraints.Ordered](want T) func(T) Ordering {
	return func(what T) Ordering {
		switch {
		case what < want:
			return Less
		case what > want:
			return Greater
		default:
			return Equal
		}
	}
}

func binary_search[T any](values []T, compare_fn func(T) Ordering) int {

	var low int = 0
	var high int = len(values)

	for low < high {

		var mid int = low + (high-low)/2
		var value T = values[mid]

		switch compare_fn(value) {
		case Greater:
			high = mid
		case Less:
			low = mid + 1
		case Equal:
			return mid
		}
	}
	return low

}

// func main() {
//
// 	var values []int = make([]int, 10)
//
// 	for i := range 10 {
// 		values[i] = i
// 	}
//
// 	println(binary_search(values, compare_function(10)))
//
// }

File: rust/src/binary_search.rs
#[derive(Debug)]
enum Ordering {
    Less,
    Greater,
    Equal,
}

fn compare_fn<T>(want: T) -> fn(T) -> Ordering {
    move |what: T| {
        if what < want {
            Ordering::Less
        } else if what > want {
            Ordering::Greater
        } else {
            Ordering::Equal
        }
    }
}

fn binary_search<T>(values: &[T], compare_fn: fn(T) -> Ordering) -> usize {
    let mut low = 0;
    let mut high = values.len();

    while (low < high) {
        let mid = low + (high - low) / 2;
        let value = values[mid];

        match compare_fn(value) {
            Ordering::Less => low = mid + 1,
            Ordering::Greater => high = mid,
            Ordering::Equal => return mid,
        }
    }
    low
}
